New lazy vals implementation

szymon-rd opened the pull request:
Cleaned up version of https://github.com/lampepfl/dotty/pull/14545

Implementing the new lazy vals scheme mentioned in https://github.com/lampepfl/dotty/pull/6979, fixing https://github.com/lampepfl/dotty/issues/7140

New PR to avoid force pushing to main branch of the previous author.

Fixes https://github.com/lampepfl/dotty/issues/15149 as well
Applied https://github.com/lampepfl/dotty/pull/14780

szymon-rd commented:
@odersky Created this as a new PR with source branch at `dotty-staging `so you can edit it.

odersky commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,171 +288,218 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   * @param rhs           the right-hand side expression of the lazy val
+   * @param tp            the type of the lazy val
+   * @param offset        the offset of the field in the bitmap
+   * @param getFlag       a flag for the volatile get function
```
Why all these additional parameters? Can't they be computed in the function themselves?

odersky reviewed (commented):
It looks overall good to me but I was not able to check details for lack of time. Can someone else give it a thorough go-over? If this is going to be the new implementation of LazyVals, it would be good to get one more person to know in detail what it does.

sjrd commented on library/src/scala/runtime/LazyVals.scala:
```diff
@@ -37,6 +38,41 @@ object LazyVals {
+   * Used to indicate the state of a lazy val that is being
+   * evaluated and of which other threads await the result.
+   */
+  final class Waiting:
```
Have you considered extending `java.util.concurrent.CountDownLatch(1)`, and directly using `countDown()` (resp. `await()`) instead of `release()` (resp. `awaitRelease()`)? Supposedly, each JVM/JDK may have an implementation of `CountDownLatch` that is optimal for its optimization patterns, and so that may be faster than a custom implementation?

szymon-rd replied:
Thanks, that's a good idea.

sjrd commented on library/src/scala/runtime/LazyVals.scala:
```diff
@@ -37,6 +38,41 @@ object LazyVals {
+   * Used to indicate the state of a lazy val that has been evaluated to
+   * `null`.
+   */
+  object NULL
```
Consider using a name that is not full-caps, to follow existing conventions. Perhaps `NullValue`?

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,171 +288,210 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   * @tailrec def x: A =
+   *    _x match
+   *    case null =>
+   *        if CAS(_x, null, Evaluating) then
```
The indentation within this comment is not regular. Please make it follow 2 spaces like everywhere else.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -101,12 +116,12 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
     }
   }
 
-
-  /** Append offset fields to companion objects
-    */
+  /** 
+   * Append offset fields to companion objects.
+   */
   override def transformTemplate(template: Template)(using Context): Tree = {
     val cls = ctx.owner.asClass
-
+    
```
Spurious whitespace.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,171 +288,210 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   * @param memberDef     the transformed lazy field member definition
+   * @param claz          the class containing this lazy val field
+   * @param target        the target synthetic field
+   * @param offset        the offset of the field in the bitmap
```
"in the bitmap" seems outdated to me. IIUC, there is no bitmap anymore.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,171 +288,210 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+    val objCasFlag = Select(ref(helperModule), lazyNme.RLazyVals.objCas)
+    val methodSymbol = memberDef.symbol.asTerm
+    val rhs = memberDef.rhs
+    val discardSymb = newSymbol(methodSymbol, lazyNme.discard, Method | Synthetic, MethodType(Nil)(_ => Nil, _ => defn.UnitType))
+    val discardDef = DefDef(discardSymb, initBlock(
+      objCasFlag.appliedTo(thiz, offset, evaluating, Select(New(waiting), StdNames.nme.CONSTRUCTOR).ensureApplied)
+        :: Return(unitLiteral, discardSymb) :: Nil))
```
What is this `discard` thing? It doesn't seem to exist in the old implementation, and none of the comments mention it.

szymon-rd replied:
It is basically:
```scala
if current.isInstanceOf[Evaluating] then
  CAS(current, Evaluating, new Waiting)
```
But it does not make sense to make separate method for that. I'm looking if the original author maybe did it for a reason I don't see, but it seems that I can just inline it.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,171 +288,210 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+        :: Return(unitLiteral, discardSymb) :: Nil))
+    // if observed a null value
+    val unevaluated = {
+      // var res: AnyRef
```
```suggestion
      // var result: AnyRef
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,171 +288,210 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+      }
+      // finally block
+      val fin = If(
+        objCasFlag.appliedTo(thiz, offset, evaluating, ref(resSymb)).equal(Literal(Constant(false))), 
```
Apply the `unary_!` method on the `Boolean` instead.

sjrd reviewed (changes requested):
Overall, this looks good. I have some suggestions here and there.

What would it take to leave the old implementation available under a Java system property? In the past in Scala 2, performance-risky changes like this had a fallback on the old implementation that could be enabled with a Java system property (or a compiler flag). I think this is a situation that would benefit from such an escape hatch.

bishabosha commented on library/src/scala/runtime/LazyVals.scala:
```diff
@@ -37,6 +38,41 @@ object LazyVals {
+     * Wakes up waiting threads. Called on completion of the evaluation
+     * of lazy val's right-hand side.
+     */
+    def release(): Unit = synchronized {
```
`synchronized` is not supported by project Loom (it uses native calls so blocks OS threads), so @sjrd comment about latches is the officially recommended workaround

szymon-rd replied:
It seems that after introducing CountDownLatch, the synchronized is not required.

szymon-rd commented:
test performance please

prolativ commented:
test performance please

bishabosha commented:
test performance please

szymon-rd commented:
test performance please

allanrenucci commented:
A few years ago, I wrote micro-benchmarks for Dotty lazy val implementations: https://github.com/allanrenucci/dotty-benchmark/tree/lazy-vals-v2

You may be interested in running these benchmarks against the new implementation.

szymon-rd commented:
test performance please

allanrenucci commented:
I don't fully remember what these objects refer to. `LazyVolatile` looks very similar to [scala.runtime.LazyVals](https://github.com/lampepfl/dotty/blob/main/library/src/scala/runtime/LazyVals.scala). Not sure about `LazySimCellWithPublicBitmap `.

allanrenucci commented:
Maybe the initialised read case should come first. I suspect it is the most common case. I.e.
```scala
def x: A =
  while true do
    val current: AnyRef = _x
    if current.isInstanceOf[A] then
      return current.asInstanceOf[A]
    else ...
```

szymon-rd commented:
@sjrd Could you take a look again? I addressed everything I think. I will look a bit into the performance to see if I can change anything in single-threaded case.

odersky commented:
I think we need a decision whether the performance hit for the uncontended init case and the initialized access case is acceptable (or whether we can improve the performance). I am not so sure about this, in fact. I believe these cases mater the most. The way I think about it is that initialized access would happen about an order of magnitude more often than uncontended init, and both are several orders of magnitude more common than the contended cases.

bishabosha commented:
it would be interesting also to compare how lazy vals perform with a large number of the new virtual threads  in Jdk 19

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
````diff
@@ -274,6 +286,212 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   * is implemented using if-s. That is:
+   * 
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * def x: A =
+   *   while true do
+   *     val current: AnyRef = _x
+   *     if current == null then
+   *       if CAS(_x, null, Evaluating) then
+   *         var result: AnyRef = null
+   *         try
+   *           result = rhs
+   *           nullable = null
+   *           if result == null then result = NullValue
+   *         finally
+   *           if !CAS(_x, Evaluating, result) then
+   *             val lock = _x.asInstanceOf[Waiting]
+   *             CAS(_x, lock, result)
+   *             lock.release()
+   *     else
+   *       if current.isInstanceOf[Evaluating] then
+   *         CAS(current, Evaluating, new Waiting)
+   *       else if current.isInstanceOf[NullValue] then
+   *         return null
+   *       else if current.isInstanceOf[Waiting] then
+   *         current.asInstanceOf[Waiting].await()
+   *       else
+   *         return current.asInstanceOf[A]
+   *   end while
````
This can help, but care must be taken to only do that when the (erased) type `A` of the field is none of the supertypes of `Waiting`. That means `Object`, obviously, but also `CountDownLatch`.

szymon-rd commented:
Ok so I think I will implement a algorithm like this:
```scala
  private var _x: AnyRef = null

  def x: A =
    if !(_x == null || _x.isInstanceOf[LazyValControlState]) then
      return _x.asInstanceOf[A]
    else if x == NullValue:
      return null
    else 
      return x_compute();
  
  private def x_compute() = 
     while true do
       val current: AnyRef = _x
       if current == null then
         if CAS(_x, null, Evaluating) then
           var result: AnyRef = null
           try
             result = rhs
             return result
           finally
              if result == null then result = NullValue
              if !CAS(_x, Evaluating, result) then
                val lock = _x.asInstanceOf[Waiting]
                CAS(_x, lock, result)
                lock.release()
       else
         if current.isInstanceOf[LazyValControlState]:
          if current.isInstanceOf[Evaluating] then
            CAS(current, Evaluating, new Waiting)
          else if current.isInstanceOf[NullValue] then
            return null
          else if current.isInstanceOf[Waiting] then
            current.asInstanceOf[Waiting].await()
         else
           return current.asInstanceOf[A]
     end while
     ```

TheElectronWill commented:
I wonder if using a special value for `Uninitialized` would be more efficient? It removes some some branches.
(this might have been suggested before, but I don't remember where)

```scala
 private var _x: AnyRef = Uninitialized // subclass of LazyValControlState

  def x: A =
    if !_x.isInstanceOf[LazyValControlState] then
      return _x.asInstanceOf[A]
    else 
      return x_compute()
  
  private def x_compute() = 
     while true do
       val current: AnyRef = _x
       if current == Uninitialized then
         if CAS(_x, Uninitialized, Evaluating) then
           var result: AnyRef = Uninitialized
           try
             result = rhs
             return result
           finally
              if !CAS(_x, Evaluating, result) then
                val lock = _x.asInstanceOf[Waiting]
                CAS(_x, lock, result)
                lock.release()
       else
         if current.isInstanceOf[LazyValControlState] then
          if CAS(current, Evaluating, new Waiting) then
            <EmptyTree>
          else if current.isInstanceOf[Waiting] then
            current.asInstanceOf[Waiting].await()
         else
           return current.asInstanceOf[A]
     end while
```

sjrd commented:
The problem with that is that you have to actually initialize the field to `Uninitialized` in the constructor. The idea of using `null` is that we get it for free when creating the object. That said, it may be worth a try, of course.

szymon-rd commented:
I was thinking about it, but it seemed a bit problematic because it introduces additional state - null before Uninitialized. If the value is referenced in the constructor before initialisation of lazy val, then we would have a problem, for example:

Right now the result in this snippet:
```
object A {
 val x = y
 lazy val y = "foo"
}
```
The x is equal to "foo". If we make it Uninitialized and treat null as null, we face two problems:
 * Firstly, the value would be equal to null, as in the constructor the `val x = y` would be called before assigning Uninitialized to the `y`. And the `null` in y would be interpreted as a null value.
 *  Secondly, it poses a problem in e.g. integer values. If in the example above we had a `lazy val y = 3`, I believe we would get a runtime error for assigning `null` to val of type `Int`. I think we would need to handle these cases separately which would cause us to add additional if-statement to the getter anyway.

Nevertheless, I think that performance-wise it's cheaper to initialize the lazy val in constructor once than to have additional if-statement on every call. However, if we do that, we are still left with the problems above.

sjrd commented:
You can "early"-initialize it, before calling the super constructor. Before that, no code can possibly refer to it.

szymon-rd commented:
I am not sure if I understand. If I have a snippet like this:
```scala
object A {
 val x = y
 lazy val y = "foo"
}
```
Then I think it would be compiled to something like this:
```scala
object A {
  val x = null
  lazy val y = null

  <init> {
      x = y() // -> null
      y = Unitialized
   }

   def y() = { //simplified
     if y == null then
       return null
    else ...
  }
}
```
It seems that the only way to fix it is to move y initialization up in the construcutor, but then it becomes weird when there are many lazy vals. I may be missing something important here.

szymon-rd commented:
Oh, ok, now I see. If I put all the `lazy = Uninitialized` as first instructions in the constructor, then it will work.

SethTisue commented:
> What would it take to leave the old implementation available under a Java system property?

Especially if we do this — but even if we don’t — I think this is a PR that would especially benefit from some good user-facing documentation — in one place, not scattered around multiple issues and PRs — of what the possible impacts on end users might be. (For example, https://github.com/lampepfl/dotty/pull/6979 has a nice pros-and-cons list, but I'm not sure how up-to-date it still is, and I had to poke around for a while to find it.)

szymon-rd commented:
After all the discussion, the schema is actually coming out differently in the end, so we will surely need to document that. I agree that it would be beneficial to add this documentation covering the aspects you mentioned. I can add it to the description of this PR after we land on the final schema that works and is efficient. Did you think about some other place where it could be a good idea to put the documentation?

szymon-rd commented:
Another interesting problem is that given a case like this:
```scala
class Foo {
  lazy val x: Any = 2;
  val y: Int = x
}
class Bar extends Foo {
  lazy val x: Any = 2
}
def main (args: Array[String]): Unit = {
  Console.print((new Bar).y);
}
```
When we use the new implementation with Uninitialized state, then we will end up with code like this:
```scala
object Uninitialized

class Foo {
  private var x_lzy: Any = null
  x_lzy = Uninitialized

  val y = x()

  def x() = {
    if(x_lzy == Uninitialized)
      x_lzy = 2
      x_lzy
    else x_lzy
  }
  
}

class Bar extends Foo {
  private var x_lzy2: Any = null
  x_lzy2 = Uninitialized

  override def x() = {
    if(x_lzy2 == Uninitialized)
      x_lzy2 = 2
      x_lzy2
    else x_lzy2
  }
}
```
And after calling `new Bar().y`, we will get null, but we should be getting 2. It makes sense, because early init is only for the scope of current class, and in the example above we have an extending class that overrides our lazy val getter, and operates on a different private var for storing the state. And this var has not been initialised yet. The way to solve it would be to operate on one var in structure, but then it couldn't be private. Another way that is a bit simpler I think could be to create a method `initUninitialized` that would put `Uninitialzied` in all the vals, and call the `super.initUninitialized` at the end. Then, when we override it and put it always first in the constructor, we would have the working early init. However, both solutions result in growing complexity that is a cost for us. @TheElectronWill do you have any thoughts on that?

TheElectronWill commented:
Do you mean
```scala
class Foo:
  private lazy val x = 2
  val y = x
class Bar:
  override lazy val x = 2
```
?
That's an interesting problem, indeed!
First of all, what do the benchmarks show for the "uninitialized" scheme? If there's no improvement it might not be worth the additional complexity.

szymon-rd commented:
Interesting note, it appears that we are trying to fix a bug that is present even with standard vals:
```scala
class Foo {
    val a = "foo"
    val b = a
}

class Bar extends Foo {
    override val a = "bar"
}

val s: String = new Bar().b
println(s) // -> null
```

But it works with current lazy vals:
```scala
class Foo {
    lazy val a = "foo"
    val b = a
}

class Bar extends Foo {
    override lazy val a = "bar"
}

val s: String = new Bar().b
println(s) // -> bar
```
So we need to fix it anyways.

sjrd commented:
Yes, that's expected that it breaks for `val`s, and I believe the initialization checker would flag it. But it's also expected that it *works* for `lazy val`s.

odersky commented:
Note: The behavior with standard vals as well as lazy vals is as expected (and, yes, it has to be maintained).

szymon-rd commented:
# New Implementation Benchmarks
I optimized the implementation in several places and here are the current results. The benchmarks were run on the `openjdk 17`, with some of them reproduced on `openjdk 11`, `openjdk 18`, `corretto`, and `graal` (to look for implementation-specific differences).

## Initialized access performance
`Initialized access benchmark` is measuring the access time to a lazy val that was initialized beforehand. 

In most cases, there is a consistent improvement of about 5% in the access time. **However**, there is a regression in one of the edge cases. When these two conditions are true: 1) the type of `lazy val` is `Any` or is erased to `Any`; 2) It is not `GraalVM` - then the performance is significantly slower. It is important to take into account that  we are talking at the order of nanoseconds, so the performance hit can be counted in handful of CPU cycles. I run the JVM in the mode that outputs the post-jit assembly code and I found that the performance hit is reflected in the number of assembly instructions, but in this case we are not hitting some kind of optimization that is hard to force-trigger.  This hit is not present on GraalVM - the `Any` case is there 5% faster than the older implementation just like in other cases.

![image](https://user-images.githubusercontent.com/4761866/191304375-2b3de45a-ac3b-4da0-8462-f24b9b8e6413.png)
(Score = ns)

## Uninitialized access performance
`Uninitialized access benchmark` is measuring the access time to a lazy val that was not initialized beforehand.

JMH Visualizer reports that the new implementation is 2% faster, but, accounting for the error threshold, we can't say that the performance is better. Therefore, I think we can conclude that it stayed roughly unchanged.

![image](https://user-images.githubusercontent.com/4761866/191312069-d7c89d8c-f7a5-4177-bae2-8bd0272c9c3c.png)


## Contended access performance
`Contended access benchmark` is measuring the access time of multiple parallel threads to an uninitialized lazy val.

There is a significant performance gain in multi-threaded cases that is growing with the number of threads.

![image](https://user-images.githubusercontent.com/4761866/191309676-2523465d-f72e-4173-aede-919f05b88fff.png)

## Summary
 - Performance in the multithreaded case is improved significantly and the general single-threaded initialized access time is improved slightly. I think it is good news, especially that performance was not the main focus of changes in this PR.
 - However, sadly, there is an edge-case on `Any`-typed lazy val, where the solution is 4x slower (+20ns). We should discuss what we can do with that.

CC @odersky @sjrd

szymon-rd commented:
test performance please

bishabosha commented:
@szymon-rd please could you fix the conflict in MiMaFilters, its breaking the benchmarks

szymon-rd commented:
It's pretty much ready for review. The feature is made experimental and I'm hoping for a release in 3.2.2. There are that many changes only because I duplicated many lazy vals tests to run them with a flag enabling this implementation.

szymon-rd commented:
@sjrd Could you take a look? I feel that it is ready for review now.

sjrd commented on tests/printing/transformed/lazy-vals-new.check:
```diff
@@ -0,0 +1,106 @@
+                    if resultNullable.==(null) then 
+                      result = scala.runtime.LazyVals.NullValue
+                     else result = resultNullable
+                    return resultNullable
```
Putting this `return resultNullable` *after* the `try..finally` may produce better bytecode. `return`s from the body of a `try` cause minutiae in the compiler back-end to be correctly handled.

sjrd commented on tests/printing/transformed/lazy-vals-new.check:
```diff
@@ -0,0 +1,106 @@
+              if current.==(scala.runtime.LazyVals.Evaluating) then 
+                {
+                  scala.runtime.LazyVals.objCAS(classOf[A], A.OFFSET$_m_0, 
+                    scala.runtime.LazyVals.Evaluating
```
```suggestion
                    current
```
to avoid re-loading `Evaluating`?

sjrd replied:
`current` is local val in this method. How could it be modified by another thread?

To be clear: I'm not suggesting to get rid of the `CAS`. Only to replace `LazyVals.Evaluating` by `current` because we already know `current eq Evaluating`.

sjrd commented on tests/printing/transformed/lazy-vals-new.check:
```diff
@@ -0,0 +1,106 @@
+                 else 
+                  if current.==(scala.runtime.LazyVals.NullValue) then 
+                    return null
+                   else ()
```
It can't be anything else than `Evaluating`, `Waiting` or `NullValue`, can it? If not, consider removing the `if current == NullValue` and directly do `return null` instead.

sjrd commented on tests/printing/transformed/lazy-vals-new.check:
```diff
@@ -0,0 +1,106 @@
+                scala.runtime.LazyVals.LazyVals$LazyValControlState
+              ]
+             then 
+              if current.==(scala.runtime.LazyVals.Evaluating) then 
```
`==` will produce bad code in this case. Use `eq` instead. Same for most of the `==` (if not all).

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -445,26 +721,37 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
     import Names.TermName
     object RLazyVals {
       import scala.runtime.LazyVals.{Names => N}
+      val waiting: TermName                = N.waiting.toTermName
```
As mentioned in the runtime lib, these should be compiler-only knowledge, so
```suggestion
      val waiting: TermName                = "waiting".toTermName
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -2,31 +2,39 @@ package dotty.tools.dotc
+   * to companion object definitions, and number of bits currently used. 
+   */
+  class LegacyOffsetInfo(defs: List[Tree], var ord: Int) extends OffsetInfo(defs)
   private val appendOffsetDefs = mutable.Map.empty[Symbol, OffsetInfo]
+  private val oldAppendOffsetDefs = mutable.Map.empty[Symbol, LegacyOffsetInfo]
```
Have you considered keeping a unique `OffsetInfo(defs, ord)` structure, and ignoring `ord`(set it to `0` and not read it) in the new codegen? This would avoid these two maps, plus paying attention to select the right one in `transformTemplate`.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -52,6 +60,11 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
     else nullables.toList
   }
 
+  private def needsBoxing(tp: Type)(using Context): Boolean = tp != NoType && tp != defn.UnitType && tp.classSymbol.isPrimitiveValueClass
```
```suggestion
  private def needsBoxing(tp: Type)(using Context): Boolean = tp != NoType && tp.classSymbol.isPrimitiveValueClass
```
`Unit` is covered in `isPrimitiveValueClass`.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -52,6 +60,11 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
 
+  private def needsBoxing(tp: Type)(using Context): Boolean = tp != NoType && tp != defn.UnitType && tp.classSymbol.isPrimitiveValueClass
+    
+  private def boxIfCan(tp: Type)(using Context): Type =
+    assert(needsBoxing(tp))
+    defn.boxedType(tp)
```
This is used only once, in a context where `needsBoxing(tp)` is closely known to be `true`. Consider inlining `defn.boxedType(tp)` directly at call site.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -101,30 +113,28 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
-
-  /** Append offset fields to companion objects
-    */
+  /** 
+   * Append offset fields to companion objects.
```
There's a bunch of spurious changes in this area. Consider reverting them to have a smaller diff and better preserve `git blame`.

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +284,274 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   *              lock.release()
+   *     else
+   *       if current.isInstanceOf[LazyValControlState] then
+   *         if current.isInstanceOf[Evaluating] then // To avoid creating Waiting instance
+   *           CAS(current, Evaluating, new Waiting)
+   *         else if current.isInstanceOf[NullValue] then
+   *           return null
+   *         else if current.isInstanceOf[Waiting] then
+   *           current.asInstanceOf[Waiting].await()
```
This does not correspond to what's actually emitted anymore. (`isInstanceOf` vs `==`, or `eq`, and the order).

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +284,274 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+        ref(valueSymbol).select(defn.Any_!=).appliedTo(nullLiteral).select(nme.And).appliedTo(ref(valueSymbol).select(defn.Any_isInstanceOf).appliedToTypeTree(controlState)
+          .select(nme.UNARY_!).appliedToNone)
+      else
+        ref(valueSymbol).select(defn.Any_isInstanceOf).appliedToTypeTree(TypeTree(tp))
```
```suggestion
        ref(valueSymbol).select(defn.Any_isInstanceOf).appliedToType(tp)
```
?

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +284,274 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+    val containerTree = ValDef(containerSymbol, nullLiteral)
+
+    // create an offset for this lazy val
+    appendOffsetDefs.get(claz) match
```
The two branches of this `match` look *very* similar. Can't they be factored out?

sjrd reviewed (changes requested):
Sorry for the delay! Here is a pretty detailed review.

szymon-rd commented on library/src/scala/runtime/LazyVals.scala:
```diff
@@ -119,11 +163,20 @@ object LazyVals {
 
 
   object Names {
+    final val controlState = "LazyValControlState"
```
True. This is the approach that was used in previous implementations so I went ahead with it and just adjusted it, but I agree that it would make more sense to put it in the compiler.

szymon-rd commented on tests/printing/transformed/lazy-vals-new.check:
```diff
@@ -0,0 +1,106 @@
+          )
+        ()
+      }
+    @static @static val OFFSET$_m_0: Long = 
```
It is also in the legacy output.

szymon-rd commented on tests/printing/transformed/lazy-vals-new.check:
```diff
@@ -0,0 +1,106 @@
+      )
+    private def writeReplace(): Object = 
+      new scala.runtime.ModuleSerializationProxy(classOf[A])
+    @volatile lazy <static> var x$lzy1: Object = null
```
The symbol of the pre-transformed val wasa annotated as static and it was being forwarded. I will remove it.

szymon-rd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +284,274 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+                      thiz: Tree,
+                      helperModule: TermSymbol,
+                      runtimeModule: String)(using Context): (DefDef, DefDef) = {
+    val tp = memberDef.tpe.widen.resultType.widen
```
Did you mean 
```scala
val tp = memberDef.tpe.widenDealias.resultType.widenDealias
```
?

szymon-rd commented:
@sjrd I responded to / applied all of your comments. Could you take a look?

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +274,231 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   *   else
+   *     return x_compute() // possible unboxing applied here
+   * 
+   * private def x_compute() =
```
```suggestion
   * private def x_compute(): AnyRef =
```
so that we don't have to guess?

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +274,231 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   *           current.asInstanceOf[Waiting].await()
+   *         else return null
+   *       else
+   *         return current.asInstanceOf[A]
```
```suggestion
   *         return current
```

sjrd commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -274,6 +274,231 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+      // if CAS(_, null, Evaluating)
+      If(
+        objCasFlag.appliedTo(thiz, offset, nullLiteral, evaluating),
+        Block(ValDef(resSymb, nullLiteral) :: ValDef(resSymbNullable, nullLiteral) :: evaluate :: Nil, // var result: AnyRef = null
+          Return(ref(resSymbNullable), lazyInitMethodSymbol)), // try ... finally ...
```
The comments on the right do not match up anymore.

sjrd reviewed (changes requested):
Looks great, thanks! I only have very minor comment-related comments left.

Could you also rebase on `main` so that there is no merge commit appearing in the PR? In fact, it looks like it would be better to *squash* all the commits of this PR anyway, at this point.

szymon-rd commented:
@sjrd I applied your suggestions, squashed some of the commits (I kept the most important "pieces" but squashed all the smaller adjustments). I also rebased on the main.

SethTisue commented:
Congrats on landing this!

Is the PR description for https://github.com/lampepfl/dotty/pull/6979 still the best place for users to read about what this change is and what implications it might have for their libraries and applications?
